New 2018/2019/2020 years

aloysbaillet opened the pull request:
Changed the base image to use NVidia's cuda-centos7-devel image so that GPUs can be made available via the nvidia runtime for docker.
Refactored the common bits into a separate `ci-common` image as it takes 1.5 hour to compile clang... I expect the common image to change rather rarely (it will change in 2021 though).

All images are now built with a argument to specify the docker hub "organisation", I'm using `aswftesting` from my own Azure jobs, and I would recommend we create a 2-stages approach to automaticaly push images to `aswfstaging` when commits are merged to master, and have a manual release step to create the `aswf` images when the staging has been deemed good enough.

@tykeal would you be able to create the following Azure Pipeline variables: `dockerhub.username` (as secret if you wish), dockerhub.password` (as secret) and dockerhub.org being `aswfstaging`. This would make the jobs work on the ASWF azure builds.

danrbailey commented on build.sh:
```diff
@@ -0,0 +1,27 @@
+# Most VFX packages are not ready for python3 yet...
+if [[ ${VFXPLATFORM_VERSION} != 2020 ]]; then
+    docker build -t ${ASWF_ORG}/ci-ocio:${ASWF_VERSION} -f ci-ocio/Dockerfile ${BUILD_ARGS} .
+    docker build -t ${ASWF_ORG}/ci-openvdb:${ASWF_VERSION} -f ci-openvdb/Dockerfile ${BUILD_ARGS} .
```
I'm wondering if we could perhaps do away with ci-base and move all the dependencies up into ci-ocio/ci-openvdb (and add a ci-openexr)? Having a three-level hierarchy (ci-common -> ci-base -> ci-openexr) seems like unnecessary complexity if it doesn't change very often. Although it's mildly wasteful, at least the large build times (ie Clang) are in the one common image and from what I remember none of the VFX dependencies took very long to compile (is that still the case?).

It feels like that might be simpler to manage if one project wished to experiment with different compiler flags to Boost for example, rather than having to mess with the hierarchy first.

aloysbaillet replied:
I see your point about simplifying the image inheritance. I don't have a strong opinion either way, but here's a summary to discuss:

Build times:
ci-common (system packages, compilers): ~ 2 hours
ci-base (python, boost, tbb, other stuff): ~20m
ci-openvdb (openexr, blosc): ~10m
ci-ocio (openexr, oiio): ~10m

I have added a "patchup.sh" that allows downstream images to install anything they wish, which can be re-installing another version of an upstream to try things, I use it in the 2020 image to update cmake as it officially supports python-3 from 3.12 onwards.

aloysbaillet replied:
One other thing that I wanted to do was to also add Qt to the base images... With Qt-5.12 coming online with 2019 I already see a different build of Qt from every vendor and even different builds of Qt from the same vendor... I'd like to point them to the ci-base as the golden reference for how we need base VFX dependencies built.
I realise that Qt will add both time and size to the base image... Ideas welcome!

danrbailey replied:
I suppose Qt creates a good reason for the three-level hierarchy as you don't really want that in the common image or the top-level images.

On the other hand, I don't think any of the existing projects (openvdb, opencolorio, openexr) have a dependency on Qt though? I agree with you that I think it's nice to have an image that has everything in it, but how about:

ci-common -> ci-base
ci-common -> ci-reference
ci-common -> ci-openvdb
ci-common -> ci-ocio
ci-common -> ci-openexr

The ci-base one would contain all the non-vfx third party dependencies as it does at the moment (python, boost, tbb, etc).
The ci-reference one would then contain the full VFX reference platform (python, boost, openexr, openvdb, Qt, etc).
The project ones wouldn't contain Qt.

For each image, we would just duplicate the installation of the python, boost, tbb, etc as it's only ~20m and I think a shallower image hierarchy tree will make things simpler.

aloysbaillet replied:
opencolorio seems to have an optional dependency on Qt, maybe @michdolan can confirm?
I'm happy to go with the flatter hierarchy though, it makes sense for simplicity.

michdolan replied:
The Qt reference in OCIO is in an example python module for an old version of Mari, so nothing needing support in a CI build. 

I'm using ci-base currently in OCIO CI to test a build against "latest" dependency versions. Having an image before any project-specific dependencies are installed is useful for cases like that. That could be a ci-reference image too though. I agree having a golden reference would be a great resource in general.

aloysbaillet replied:
Thanks for the update!

danrbailey replied:
@michdolan - it would be great to get a "latest" docker image, we'd really like that too!

This is a bit of a side discussion, but one workflow issue I see with the latest builds as a native part of the CI is that the build isn't necessarily reproducible. If a PR is submitted and everything is green, then an additional commit is added with some feedback addressed, the build may pick up a newer version of a dependency and fail. This could cause confusion to the user thinking it's their commit that has broken the build.

I'm not sure the best practice to resolve this, but I'm thinking of a workflow such as having a weekly build that runs against the latest image and then have PRs run off an earlier "latest" image, one that has (hopefully) already been successful. This would all be managed by tags.

Perhaps a weekly cron could check for any latest versions of any dependency and automatically push a change to trigger a new build generating a new image if needed? Either way, I think it would be great to support latest as a native part of this docker repo...

aloysbaillet replied:
I was also thinking to add "integration tests" for these images so the `ci-openvdb` is used to build the matching vfx platform version of OpenVDB, maybe these could run before release, and would ensure that `aswf/ci-openvdb:latest` should never break opencdb builds...

michdolan replied:
Yeah @danrbailey , that's a good point about running the "latest" build as a cron vs. for PRs. I actually hit that very issue recently where I made another commit to my Azure PR and the OpenEXR master branch build failed. I modified our build so that "latest" pulls the latest tag instead of master to provide some level of stability, but ultimately we'll likely separate it as you suggested.

danrbailey commented on azure-pipelines.yml:
```diff
@@ -1,15 +1,37 @@
+  jobs:
+  - job:
+    condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
+    timeoutInMinutes: 360
```
Unless there's a good reason to use a lower limit, perhaps just set this to 0 so it uses the maximum (5 hours)?

aloysbaillet replied:
The default is one hour as far as know? (I got a few timeout without this...)

danrbailey replied:
Right, this is a bit unintuitive, but:

no timeout specified - 1 hour
timeoutInMinutes: 0 - 5 hours
timeoutInMinutes: 360 - 6 hours (perhaps this rounds down to 5 hours as that's the max though?)

So yeah, if you explicitly set the timeoutInMinutes to 0, it should just use the highest limit instead of 1 hour (as I understand it).

aloysbaillet replied:
Hehe I didn't notice that, fun!

danrbailey commented on azure-pipelines.yml:
```diff
@@ -1,15 +1,37 @@
+  condition: always()
+  jobs:
+  - job:
+    condition: eq(variables['Build.SourceBranch'], 'refs/heads/master')
```
So this job still gets triggered on any commit to master? Is there a way of committing a change to master that triggers all _but_ the ci-common image being built (in the more common case where you're only changing VFX dependencies)?

aloysbaillet replied:
I would also like another way of triggering these, a simple update to the readme will consume many hours of compute, and that makes me feel bad for the planet!
I'm thinking of using tags to trigger the pushes, and more fine grain triggers to only run relevant jobs. But that will take some more work.

danrbailey replied:
I think part of this can be done with path filters:

https://docs.microsoft.com/en-us/azure/devops/pipelines/build/triggers?view=azure-devops&tabs=yaml#paths

We can always exclude certain directories or files in the repo from the triggers.

For the other aspect of this, perhaps we could have a refs/head/common branch and a refs/head/master branch and trigger ci-common from the former and all the others from the latter?

I think I'd prefer triggering from branches than tags to be honest, just because it would be a simpler workflow, but I appreciate the desire to use tags. It's complicated further by having Git tags and Docker tags which wouldn't necessarily line up too.

michdolan replied:
Having multiple branches sounds like a good safety. You would still want a commit to refs/head/common to trigger a push of refs/head/master in that case, but that would occur much less often.

danrbailey commented on azure-pipelines.yml:
```diff
@@ -1,15 +1,37 @@
+  - template: .azure/build-linux-docker-images.yml
+    parameters:
+      vfxplatform_version: 2018
+      aswf_version: 2018
```
Is there a situation where these two parameters might be different?

aloysbaillet replied:
I think they will be once we start using tags trigger the uploads. I anticipate we would use something like 2018.0 and once we realise we need something else we can tag 2018.1 and that's what would be released... This is all a bit theoretical at this point though!

michdolan commented on ci-ocio/Dockerfile:
```diff
@@ -1,6 +1,44 @@
+    /tmp/install_python.sh && \
+    /tmp/install_boost.sh && \
+    /tmp/install_tbb.sh && \
+    /tmp/install_numpy.sh && \
```
ci-ocio needs /tmp/install_glew.sh

I think install_glew can be removed from ci-openvdb. Is that right @danrbailey ?

aloysbaillet replied:
Whoops sorry I mixed up the two, I think this is fixed now.

michdolan replied:
Thanks!

aloysbaillet commented:
I've got a good build there: https://aloysbaillet.visualstudio.com/ASWF/_build/results?buildId=190
The updated setup builds all 10 images in parallel in take only 1/2h which is pretty good!

aloysbaillet commented:
Thanks Dan!
I'd like to add one more thing to this PR which is to add a few more tags (including `latest`).

danrbailey commented:
> Thanks Dan!
> I'd like to add one more thing to this PR which is to add a few more tags (including `latest`).

I'm a bit concerned about building our CI on mutable tags. I'd like to be able to go back to a PR that might have been open for a year (for example) and the docker image not have changed underneath it in that time so that it no longer builds. It's for that reason that I'm wondering about using an explicit tag version rather than "latest", then an OpenVDB PR that adds support for VFX 2020 for example would also bump the docker image version to pick up an image that adds the relevant dependencies. This ties into how we could best support a "latest" image too. Perhaps a bigger discussion better suited for a CI working group meeting though...

aloysbaillet commented:
I've clarified the use of tags there: https://github.com/aloysbaillet/aswf-docker/tree/refactor_with_more_vfx_years#image-tags

danrbailey commented:
> I've clarified the use of tags there: https://github.com/aloysbaillet/aswf-docker/tree/refactor_with_more_vfx_years#image-tags

Ah, I must have missed that, good documenting! :)

Only minor question is whether people might confuse the second number with "months" or "quarters", ie 2019.3 meaning 2019 Q3 or USD-style 2019.6 meaning June 2019 release. Perhaps that doesn't matter, but mentioning it anyway.

aloysbaillet commented:
I just added these lines in my last commit :-)
I also just added a line to clarify that the second number is *not* a month indeed!
My last readme update didn't trigger a full rebuild, so that's that...
I wouldn't mind if you could re-approve the PR, thanks!

aloysbaillet commented:
Looks like @michdolan will have to approve as well before we can merge.
That could give @tykeal enough time to configure the ASWF Azure project with the required docker info! I'll contact him offline.

tykeal commented:
@aloysbaillet and @danrbailey I've created a variable group called dockerhub-staging that has the 3 dockerhub.???? variables defined in it. That groups should be available to all pipelines now.

Unless I was misunderstanding how the variables should be created ;)

aloysbaillet commented:
Thanks everyone!
I just added a line to use the variable group "dockerhub-staging" and I got it to work on my Azure project
On there: https://aloysbaillet.visualstudio.com/ASWF/_build/results?buildId=200&view=logs&s=fa738274-c5e4-5bf5-3478-aa610f588840&j=962732c0-e535-5d5d-6ec3-c968cb842735
I get the following expansion:
```
Script contents:
docker login -u *** -p ***
```
But on there: https://dev.azure.com/academysoftwarefoundation/Academy%20Software%20Foundation/_build/results?buildId=151&view=logs&jobId=47c4f9df-65a4-5a43-350e-30b835999330&taskId=a24a542e-d24c-5b21-5f2f-85891dfcfbb5&lineStart=10&lineEnd=11&colStart=1&colEnd=63
I get this error:
```
Script contents:
docker login -u $(dockerhub.username) -p $(dockerhub.password)
```

The variables have not been replaced... I suspect there must be a difference in the way our azure projects are configured?
@tykeal have you checked the "Enabled Multi-Stage Pipelines" in the "Preview Features" menu? https://go.microsoft.com/fwlink/?linkid=2097082

I've also fixed my DCO errors.

aloysbaillet commented:
I found that there might be another settings to change called "Make secrets available to builds of forks" (see https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#validate-contributions-from-forks ) to enable so that forks can run docker login.
That said this presents some security risks as anybody could create a fork, modify the build scripts in ways that could leak the secrets...
I'll try and find a way to avoid using `docker login` unless the branch is master, unless you have a better idea?

tykeal commented:
> The variables have not been replaced... I suspect there must be a difference in the way our azure projects are configured?
> @tykeal have you checked the "Enabled Multi-Stage Pipelines" in the "Preview Features" menu? https://go.microsoft.com/fwlink/?linkid=2097082

That was not enabled for the org, but it is now.

> I found that there might be another settings to change called "Make secrets available to builds of forks" (see https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#validate-contributions-from-forks ) to enable so that forks can run docker login.
> That said this presents some security risks as anybody could create a fork, modify the build scripts in ways that could leak the secrets...
> I'll try and find a way to avoid using `docker login` unless the branch is master, unless you have a better idea?

Yeah, I'm hesitant to do that since it does present a security risk for shared secrets.

Give it another whirl?

aloysbaillet commented:
Good news @tykeal I've managed to use `docker save/load` to save the intermediate docker image so that there's no need to `docker login` between 2 stages.
The login will only be required on the master branch for now.
Could you approve my PR so we can merge?

Oh and also I noticed that `dockerhub-staging` variable group points to `aswfstaging`, but we'd like to start populating the real `aswf` docker organisation now, should we use different variable groups for the `aswf` org?

tykeal commented:
> Oh and also I noticed that `dockerhub-staging` variable group points to `aswfstaging`, but we'd like to start populating the real `aswf` docker organisation now, should we use different variable groups for the `aswf` org?

I've created another variable group of `dockerhub-release` that has the release vars and credentials in it. I wanted to make sure that you were testing against the staging registry first ;)

aloysbaillet commented:
Thanks! I'll merge this one so we can see that the docker push works from master, then I'll open another PR to use `dockerhub-release`. I'm thinking we could use 2 different branches that use different variable groups...
